src/consumption.cpp

```diff
@@ -52,6 +55,89 @@ int player::nutrition_for( const itype *comest ) const
     return ( comest && comest->comestible ) ? comest->comestible->nutr : 0;
 }
 
+std::map<vitamin_id, int> player::vitamins_from( const itype_id &id ) const
+{
+    return vitamins_from( item( id ) );
+}
+
+std::map<vitamin_id, int> player::vitamins_from( const item &it ) const
+{
+    std::map<vitamin_id, int> res;
+
+    if( !it.type->comestible ) {
+        return res;
+    }
+
+    // food to which the player is allergic to never contains any vitamins
+    if( allergy_type( it ) != MORALE_NULL ) {
+        return res;
+    }
+
+    // @todo bionics and mutations can affect vitamin absorption
+    for( const auto &e : it.type->comestible->vitamins ) {
+        res.emplace( e.first, e.second );
+    }
+
+    return res;
+}
+
+int player::vitamin_rate( const vitamin_id &vit ) const
+{
+    int res = vit.obj().rate();
+
+    for( const auto &m : get_mutations() ) {
+        const auto &mut = mutation_branch::get( m );
+        auto iter = mut.vitamin_rates.find( vit );
+        if( iter != mut.vitamin_rates.end() ) {
+            res += iter->second;
+        }
+    }
+
+    return std::max( res, 0 );
+}
+
+int player::vitamin_mod( const vitamin_id &vit, int qty, bool capped )
+{
+    auto it = vitamin_levels.find( vit );
+    if( it == vitamin_levels.end() ) {
+        return 0;
+    }
+    const auto &v = it->first.obj();
+
+    if( qty > 0 ) {
+        // accumulations can never occur from food sources
+        it->second = std::min( it->second + qty, capped ? 0 : v.max() );
+
+    } else if( qty < 0 ) {
+        it->second = std::max( it->second + qty, v.min() );
+    }
+
+    auto eff = v.effect( it->second );
+    if( !eff.is_null() ) {
+        // consumption rate may vary so extend effect until next check due for this vitamin
+        add_effect( eff, ( vitamin_rate( vit ) * 10 ) - get_effect_dur( eff ) + 1 );
```

Comment:
That causes a possible problem: player isn't informed about food fixing deficiencies when eating, only at the next "effect hour", which ticks independently from main game timer. This means that the information about effect timing out will appear somewhere between right now and in the next hour (assuming standard rate).

Also, it's a good idea to use `MINUTES` and `HOURS` macros in effect timing. In this case the 10 being changed to `MINUTES(1)` would show the intent more clearly.